wallet/src/libwallet/transaction.rs

```diff
@@ -18,37 +18,464 @@ use rand::thread_rng;
+	/// Receiver has sender's public data, has filled their public data and
+	/// part-signed
+	ReceiverInitiation,
+	/// Sender has all public data, has filled it and signed
+	SenderConfirmation,
+	/// Reciever has all data, and has signed with their output
+	ReceiverConfirmation,
+}
+/// Public data for each participant in the slate
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct ParticipantData {
+	/// Id of participant in the transaction. (For now, 0=sender, 1=rec)
+	pub id: u64,
+	/// Public key corresponding to private blinding factor
+	pub public_blind_excess: PublicKey,
+	/// Public key corresponding to private nonce
+	pub public_nonce: PublicKey,
+	/// Public partial signature
+	pub part_sig: Option<Signature>,
+}
+
+impl ParticipantData {
+	/// A helper to return whether this paricipant
+	/// has completed round 1 and round 2;
+	/// Round 1 has to be completed before instantiation of this struct
+	/// anyhow, and for each participant consists of:
+	/// -Inputs added to transaction
+	/// -Outputs added to transaction
+	/// -Public signature nonce chosen and added
+	/// -Public contribution to blinding factor chosen and added
+	/// Round 2 can only be completed after all participants have
+	/// performed round 1, and adds:
+	/// -Part sig is filled out
+	pub fn is_complete(&self) -> bool {
+		self.part_sig.is_some()
+	}
+}
+
+/// A 'Slate' is passed around to all parties to build up all of the public
+/// tranaction data needed to create a finalised tranaction. Callers can pass
+/// the slate around by whatever means they choose, (but we can provide some
+/// binary or JSON serialisation helpers here).
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Slate {
+	/// The phase the slate is in (how much data has been filled out)
+	pub phase: SlatePhase,
+	/// The number of participants intended to take part in this transaction
+	pub num_participants: usize,
+	/// Unique transaction ID, selected by sender
+	pub id: Uuid,
+	/// The core transaction data:
+	/// inputs, outputs, kernels, kernel offset
+	pub tx: Transaction,
+	/// base amount (excluding fee)
+	pub amount: u64,
+	/// fee amount
+	pub fee: u64,
+	/// Block height for the transaction
+	pub height: u64,
+	/// Lock height
+	pub lock_height: u64,
+	/// Participant data, each participant in the transaction will
+	/// insert their public data here. For now, 0 is sender and 1
+	/// is receiver, though this will change for multi-party
+	pub participant_data: Vec<ParticipantData>,
+}
+
+impl Slate {
+	/// Create a new slate
+	pub fn blank(num_participants: usize) -> Slate {
+		Slate {
+			phase: SlatePhase::SenderInitiation,
+			num_participants: num_participants,
+			id: Uuid::new_v4(),
+			tx: Transaction::empty(),
+			amount: 0,
+			fee: 0,
+			height: 0,
+			lock_height: 0,
+			participant_data: vec![],
+		}
+	}
+
+	/// Adds selected inputs and outputs to the slate's transaction
+	/// Returns blinding factor
+	pub fn add_transaction_elements(
```

Comment:
This is getting closer now to what I think it should look like... all parties involved add their inputs/outputs to the transaction, then all parties have to perform round 1, then all parties have to perform round 2 (using the provided functions). This workflow should scale easily to support multi-party/multi-sig scenarios.